### PR TITLE
JBS-436: Support resizing of volumes after a restore of snapshot

### PR DESCRIPTION
**Pull request by shishirng**
If the dst volume size > snapshot size (src volume size from which
snap was taken), the volume gets shrunk to the src size after a
volume restore (import-diff). We need to explicitly call image
resize if dst volume size is > than snapshot size.

Signed-off-by: shishir gowda <shishir.gowda@ril.com>

### CODE REVIEW

**Comment by shishirng** [comment]
Test case:
==Snapshot==
    <item>
      <status>completed</status>
      <snapshotId>5087d65a-f61c-4043-a341-50d201829a22</snapshotId>
      <volumeSize>8</volumeSize> <=====8GB size
      <volumeId>eb6190f5-5380-487d-ab17-78cf4ecf2765</volumeId>
      <startTime>2016-06-27T11:56:19.000000</startTime>
    </item>


==restore to Same size==
    <item>
      <status>available</status>
      <encrypted>false</encrypted>
      <volumeType>standard</volumeType>
      <volumeId>8bf926fa-fd26-4a07-ab25-a46dfbced50a</volumeId>
      <attachmentSet/>
      <snapshotId>5087d65a-f61c-4043-a341-50d201829a22</snapshotId>
      <createTime>2016-06-27T12:04:29.000000</createTime>
      <size>8</size>
    </item>


sudo rbd info sbs/volume-8bf926fa-fd26-4a07-ab25-a46dfbced50a
rbd image 'volume-8bf926fa-fd26-4a07-ab25-a46dfbced50a':
	size 8192 MB in 2048 objects <====8GB
	order 22 (4096 kB objects)
	block_name_prefix: rbd_data.3c02a64ae5da12
	format: 2
	features: layering, striping
	flags: 
	stripe unit: 4096 kB
	stripe count: 1


==Restore same snapshot to larger size - 10GB ==
    <item>
      <status>available</status>
      <encrypted>false</encrypted>
      <volumeType>standard</volumeType>
      <volumeId>3e07173f-a1a6-4176-8d99-d7241acfb691</volumeId>
      <attachmentSet/>
      <snapshotId>5087d65a-f61c-4043-a341-50d201829a22</snapshotId>
      <createTime>2016-06-27T12:14:46.000000</createTime>
      <size>10</size>
    </item>

sudo rbd info sbs/volume-3e07173f-a1a6-4176-8d99-d7241acfb691
rbd image 'volume-3e07173f-a1a6-4176-8d99-d7241acfb691':
	size 10240 MB in 2560 objects <===10GB now correctly reflecting size
	order 22 (4096 kB objects)
	block_name_prefix: rbd_data.3c062a729fdf63
	format: 2
	features: layering, striping
	flags: 
	stripe unit: 4096 kB
	stripe count: 1

cinder-backup log:

2016-06-27 12:14:51.249 55828 INFO cinder.backup.drivers.sbs [req-c0ac5d63-78 0c93daf7391c46f99474cf5b0ca7a0e4 00000000000000000
000631677221232 - - -] Expanding volume 3e07173f-a1a6-4176-8d99-d7241acfb691 from 8GB to 10GB

File: cinder/backup/drivers/sbs.py [hunk 1]
```diff
@@ -761,6 +761,19 @@ def _download_from_DSS(self, snap_name, volume_name, ceph_args, container, backu
...
         return
 
+    def _restore_resize_volume(self, volume_name, new_size):
+        with rbd_driver.RADOSClient(self, self._ceph_backup_pool) as client:
+            volume_rbd = None
+            volume_rbd = self.rbd.Image(client.ioctx,
+                                        encodeutils.safe_encode(volume_name),
+                                        read_only=False)
+            try:
+                volume_rbd.resize(new_size)
```
Comment (aggarwalchirag) [inline cinder/backup/drivers/sbs.py]: I think we should have exception handling here: catch the exception, print the log and raise some known exception type

File: cinder/backup/drivers/sbs.py [hunk 2]
```diff
@@ -781,6 +794,19 @@ def _restore_rbd(self, backup, volume_id, volume_file, ceph_args):
...
                                 backup['container'], backup)
+
+        # If volume size is > size of snapshot volume, then after a restore
+        # the volume gets shrunk to size of snapshot volume. We need to call
+        # resize of image to expand volume after a restore. Pass length which
+        # is size * GB units
+
+        volume_size = int(volume['size'])
+        backup_size = int(backup['size'])
+        if volume_size > backup_size:
```
Comment (aggarwalchirag) [inline cinder/backup/drivers/sbs.py]: what if we are restoring using incremental backups? Does backup_size give us the right size in that case ?
Also, will we be resizing after each incremental restore ?

File: cinder/backup/drivers/sbs.py [hunk 3]
```diff
@@ -781,6 +794,19 @@ def _restore_rbd(self, backup, volume_id, volume_file, ceph_args):
...
+            LOG.info("Expanding volume %s from %dGB to %dGB" %
+                     (volume_id, backup_size, volume_size))
+            self._restore_resize_volume(volume_name, length)
+
```
Comment (aggarwalchirag) [inline cinder/backup/drivers/sbs.py]: I think this should be part of _restore_v1 and not _restore_rbd

File: cinder/backup/drivers/sbs.py (hunk 1 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: ack. Will do that

File: cinder/backup/drivers/sbs.py (hunk 3 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: My bad. Will move it to _restore_v1

File: cinder/backup/drivers/sbs.py (hunk 2 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: backup size always gives the original volume size, so we are good here.

But we have to move this to be done after all the the restores are done.

File: cinder/backup/drivers/sbs.py [hunk 4]
```diff
@@ -781,6 +781,39 @@ def _restore_rbd(self, backup, volume_id, volume_file, ceph_args):
 
         self._download_from_DSS(backup_name, volume_name, ceph_args,
                                 backup['container'], backup)
+
+        return
+
+    def _restore_resize_volume(self, backup, volume_id):
+        backup_id = backup['id']
+        volume = self.db.volume_get(self.context,volume_id)
+        length = int(volume['size']) * units.Gi
```
Comment (aggarwalchirag) [inline cinder/backup/drivers/sbs.py]: i am assuming that there will be no overflow here. Please check once.

**Comment by shishirng** [comment]
==8GB snapshot==


 \<CreateSnapshotResponse xmlns="http://compute.jiocloudservices.com/doc/2016-03-01/">
  \<requestId>req-c0ac5d63-78</requestId>
  \<status>pending</status>
  \<snapshotId>98d4860f-e809-403b-82f5-81660b04c149</snapshotId>
  \<volumeSize>8</volumeSize>
  \<volumeId>eb6190f5-5380-487d-ab17-78cf4ecf2765</volumeId>
  \<startTime>2016-06-28T13:36:12.000000</startTime>
\</CreateSnapshotResponse>
==10G volume==
\<CreateVolumeResponse xmlns="http://compute.jiocloudservices.com/doc/2016-03-01/">
  \<requestId>req-c0ac5d63-78</requestId>
  \<status>creating</status>
  \<encrypted>false</encrypted>
  \<volumeType>standard</volumeType>
  \<volumeId>df4a1cff-2d25-494c-a6cb-4177b061c4dc</volumeId>
  \<attachmentSet/>
  \<snapshotId>98d4860f-e809-403b-82f5-81660b04c149</snapshotId>
  \<createTime>2016-06-28T14:26:40.000000</createTime>
  \<size>10</size>
\</CreateVolumeResponse>

2016-06-28 14:26:45.462 11211 INFO cinder.backup.drivers.sbs [req-c0ac5d63-78 0c93daf7391c46f99474cf5b0ca7a0e4 00000000000000000000631677221232 - - -] Expanding volume df4a
1cff-2d25-494c-a6cb-4177b061c4dc from 8GB to 10GB


==8 GB volume ==
\<CreateVolumeResponse xmlns="http://compute.jiocloudservices.com/doc/2016-03-01/">
  \<requestId>req-c0ac5d63-78</requestId>
  \<status>creating</status>
  \<encrypted>false</encrypted>
  \<volumeType>standard</volumeType>
  \<volumeId>e7fbd3ba-b45d-424e-b8a4-db28d42cf9ad</volumeId>
  \<attachmentSet/>
  \<snapshotId>98d4860f-e809-403b-82f5-81660b04c149</snapshotId>
  \<createTime>2016-06-28T14:30:08.000000</createTime>
  \<size>8</size>
\</CreateVolumeResponse>

sudo rbd info sbs/volume-e7fbd3ba-b45d-424e-b8a4-db28d42cf9ad
rbd image 'volume-e7fbd3ba-b45d-424e-b8a4-db28d42cf9ad':
	size 8192 MB in 2048 objects
	order 22 (4096 kB objects)
	block_name_prefix: rbd_data.3dec8d4b6b0a81
	format: 2
	features: layering, striping
	flags: 
	stripe unit: 4096 kB
	stripe count: 1
==

File: cinder/backup/drivers/sbs.py [hunk 5]
```diff
@@ -822,6 +855,9 @@ def _restore_v1(self, backup, volume_id, volume_file):
             LOG.error(_LE('Restore to volume %(volume)s finished with error - '
                           '%(error)s.'), {'error': e, 'volume': volume_id})
             raise
+
+        self._restore_resize_volume(backup, volume_id)
```
Comment (aggarwalchirag) [inline cinder/backup/drivers/sbs.py]: i think, this call should be made before calling fsync

File: cinder/backup/drivers/sbs.py (hunk 4 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: Ack, integer will be 19bytes of size. With that we should be able to store upto 999TB

File: cinder/backup/drivers/sbs.py (hunk 5 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: Ack. Will make the changes

**Pull request merged**